haystack/dataclasses/chat_message.py

```diff
@@ -5,104 +5,306 @@
+
+LEGACY_INIT_PARAMETERS = {"role", "content", "meta", "name"}
 
 
 class ChatRole(str, Enum):
-    """Enumeration representing the roles within a chat."""
+    """
+    Enumeration representing the roles within a chat.
+    """
 
-    ASSISTANT = "assistant"
+    #: The user role. A message from the user contains only text.
     USER = "user"
+
+    #: The system role. A message from the system contains only text.
     SYSTEM = "system"
+
+    #: The assistant role. A message from the assistant can contain text and Tool calls. It can also store metadata.
+    ASSISTANT = "assistant"
+
+    #: The tool role. A message from a tool contains the result of a Tool invocation.
+    TOOL = "tool"
+
+    #: The function role. Deprecated in favor of `TOOL`.
     FUNCTION = "function"
 
+    @staticmethod
+    def from_str(string: str) -> "ChatRole":
+        """
+        Convert a string to a ChatRole enum.
+        """
+        enum_map = {e.value: e for e in ChatRole}
+        role = enum_map.get(string)
+        if role is None:
+            msg = f"Unknown chat role '{string}'. Supported roles are: {list(enum_map.keys())}"
+            raise ValueError(msg)
+        return role
+
+
+@dataclass
+class ToolCall:
+    """
+    Represents a Tool call prepared by the model, usually contained in an assistant message.
+
+    :param id: The ID of the Tool call.
+    :param tool_name: The name of the Tool to call.
+    :param arguments: The arguments to call the Tool with.
+    """
+
+    tool_name: str
+    arguments: Dict[str, Any]
+    id: Optional[str] = None  # noqa: A003
+
+
+@dataclass
+class ToolCallResult:
+    """
+    Represents the result of a Tool invocation.
+
+    :param result: The result of the Tool invocation.
+    :param origin: The Tool call that produced this result.
+    :param error: Whether the Tool invocation resulted in an error.
+    """
+
+    result: str
+    origin: ToolCall
+    error: bool
+
+
+@dataclass
+class TextContent:
+    """
+    The textual content of a chat message.
+
+    :param text: The text content of the message.
+    """
+
+    text: str
+
+
+ChatMessageContentT = Union[TextContent, ToolCall, ToolCallResult]
+
 
 @dataclass
 class ChatMessage:
     """
     Represents a message in a LLM chat conversation.
 
-    :param content: The text content of the message.
-    :param role: The role of the entity sending the message.
-    :param name: The name of the function being called (only applicable for role FUNCTION).
-    :param meta: Additional metadata associated with the message.
+    Use the `from_assistant`, `from_user`, `from_system`, and `from_tool` class methods to create a ChatMessage.
     """
 
-    content: str
-    role: ChatRole
-    name: Optional[str]
-    meta: Dict[str, Any] = field(default_factory=dict, hash=False)
+    _role: ChatRole
+    _content: Sequence[ChatMessageContentT]
+    _meta: Dict[str, Any] = field(default_factory=dict, hash=False)
```

Comment:
I'd say we still need `name` for some multi-user/multi-agent scenarios (see https://platform.openai.com/docs/api-reference/chat). Shall we add it back and give it a default value of `None`?

Reply:
What do you refer to @LastRemote - I think I added name, a year ago almost, for some early function calling support but the tooling structured output has changed a few times since then. Where exactly is name needed now as I could not find any references to it in the link above 🤔

Reply:
> What do you refer to @LastRemote - I think I added name, a year ago almost, for some early function calling support but the tooling structured output has changed a few times since then. Where exactly is name needed now as I could not find any references to it in the link above 🤔

It's an optional part of message object. So you can have something like this in an OpenAI request:
```
[
  {"name": "Joe", "role": "assistant", "content": "Hi, I'm Joe!"},
  {"name": "Alice", "role": "assistant", "content": "Hi, I'm Alice!"}
]
```
<img width="599" alt="Screenshot 2024-12-17 at 16 51 57" src="https://github.com/user-attachments/assets/a2a53426-86e5-491f-aca1-6f9205dc8894" />

Reply:
Aha, it was collapsed so it didn't show up in the search. You’re right! In a multi-agent environment one can have responses from multiple LLMs so it could be useful to tag which message is coming from whom. @anakin87 wdyt?

Reply:
I'm taking a look...
It is only supported by OpenAI and only in user/assistant/system message, not in tool messages.

Reply:
I readded this field in https://github.com/deepset-ai/haystack/pull/8640/commits/328cebd98696257041933c4973da20157f8d4812